### DISCUSSION TITLE
Q: Failing to detect match

### DISCUSSION THREAD

**Discussion by Jmoore1127** [alecthomas/participle#285] [Q&A] [answered]
I am attempting to parse a very simple DSL. I am running into an error in my testing but can't determine if my grammar is wrong, I need a different lexer, or there is an actual bug somewhere. I assume the former since this is fairly straight forward.

A basic example

```golang
type RelNode struct {
  Default bool `@'-->'?`
  Reltype string `| ('-'@Ident'->')
}
```

This works 
```golang
parser.ParseString("", "-COMPOSED->")
// RelNode {
//   Default: false,
//   Reltype: "COMPOSED",
// }
```

But this fails with an error
```golang
parser.ParseString("", "-->")
```
`panic: 1:1: branch "-->"? was accepted but did not progress the lexer at 1:1 ("-") [recovered]`

I am not sure I understand the error, if it matched (was accepted?) why wouldn't the lexer advance?

Edit:
I've tried various other tweaks which usually result in a `token '-' not expected` error 
e.g 
```golang
type RelNode struct {
	Default bool   `(@'-->'`
	Reltype string ` | '-'@Ident'->')`
}

//Other groupings / optionals

```

**Comment by alecthomas** [comment, marked as answer]
Assuming you're using the default lexer, this is almost certainly because that lexer does not return `-->` as a single token, it will return `['-', '-', '>']`. You'll either need a custom lexer, or to accept each token individually in your grammar.

**Reply by Jmoore1127** [reply]
Thank you for the quick reply. Converting to `@"-""-"">"?` fixes the issue. Re-reading the Grammar section I see the call-out about matching tokens exactly.
